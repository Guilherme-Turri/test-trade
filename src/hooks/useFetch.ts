import React from 'react'

export const useFetchData = () => {
  const [error, setError] = React.useState<string | null>(null);
  const [loading, setLoading] = React.useState<boolean>(false);

  const request = React.useCallback(async <T> (url: string, options: RequestInit) => {
    let response;
    let json:T | null;
    setLoading(true)
    try {
      setError(null)
      response = await fetch(url, options)
      json = await response.json()
      if (!response.ok) throw new Error()
    } catch (e: any) {
      json = null;
      setError('Falha ao conectar com o servidor.')
    }
    finally {
      setLoading(false)
          }
    return { response, json } 
  }, []);
  return { error, request, loading }
} 