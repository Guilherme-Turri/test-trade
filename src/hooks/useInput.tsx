import React from 'react'

const typeField = {
  email: {
    regex: /^.{2,}$/,
    message: "Por favor, digite duas letras ou mais."
  }
}

export const useInput = () => {
  const [value, setValue] = React.useState<string>('')
  const [error, setError] = React.useState<string>('');

  const validate = (value: string) => {
    if (typeField.email.regex.test(value) === false) setError(typeField.email.message)
    else setError('');
  }

  const onChange = ({ target }: React.ChangeEvent<HTMLInputElement>) => {
    setValue(target.value)
    if (target.placeholder === "Nome") {
      validate(target.value)
    }
  }

  return (
    { value, onChange, error, setValue }
  )
}