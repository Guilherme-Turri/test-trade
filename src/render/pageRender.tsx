import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../styles/theme';
import store from '../storage/store';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';

export const pageRender = (children: React.ReactNode) => {
  return render(
    <ThemeProvider theme={theme}>
      <Provider store={store}>
        <BrowserRouter>
          {children}
        </BrowserRouter>
      </Provider>
    </ThemeProvider>);
};