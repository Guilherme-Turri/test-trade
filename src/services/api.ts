const API_URL = ' https://v3.football.api-sports.io/'

export function AUTH_USER_API(apiKey:string) {
  return {
    url: `${API_URL}countries`,
    options: {
      method: 'GET',
      headers: {
      'x-rapidapi-host': 'v3.football.api-sports.io',
      'x-rapidapi-key': apiKey
    },
    },
  };
}
export function SEASON_API (apiKey:string) {
  return {
    url: `${API_URL}leagues/seasons`,
    options: {
      method: 'GET',
      headers: {
      'x-rapidapi-host': 'v3.football.api-sports.io',
      'x-rapidapi-key': apiKey
    },
    },
  };
}
export function LEAGUES_API (apiKey:string, country:string, season:string) {
    return {
    url: `${API_URL}leagues?country=${country}&season=${season}`,
    options: {
      method: 'GET',
      headers: {
      'x-rapidapi-host': 'v3.football.api-sports.io',
      'x-rapidapi-key': apiKey
    },
    },
  };
}
export function CLUB_API (apiKey:string, league:string, season:string) {
  return {
    url: `${API_URL}teams?league=${league}&season=${season}`,
    options: {
      method: 'GET',
      headers: {
      'x-rapidapi-host': 'v3.football.api-sports.io',
      'x-rapidapi-key': apiKey
    },
    },
  };
}
export function STATISTICS_API (apiKey:string, league:string, season:string, club:string) {
    return {
    url: `${API_URL}teams/statistics?league=${league}&season=${season}&team=${club}`,
    options: {
      method: 'GET',
      headers: {
      'x-rapidapi-host': 'v3.football.api-sports.io',
      'x-rapidapi-key': apiKey
    },
    },
  };
}

export function CHART_STATISTICS_API (apiKey:string, league:string, season:string, club:string) {
  return {
  url: `${API_URL}players?league=${league}&season=${season}&team=${club}`,
  options: {
    method: 'GET',
    headers: {
    'x-rapidapi-host': 'v3.football.api-sports.io',
    'x-rapidapi-key': apiKey
  },
  },
};
}