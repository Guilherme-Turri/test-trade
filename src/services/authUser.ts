import { AUTH_USER_API } from "./api";
import { IRequest, ICountry } from "../types/types";

interface AuthCountries{
  response?:ICountry[]
  errors?:{
    token:string
  }
}

export const authUser = async (apiKey:string , request:IRequest<AuthCountries>)=> {
  const { url, options } = AUTH_USER_API(apiKey)
  const {response, json} = await request(url, options)
  if(response && json){
    return {response, json}
   }
    else return 
 }
 