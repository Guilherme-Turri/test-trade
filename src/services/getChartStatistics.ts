import { CHART_STATISTICS_API } from "./api";
import { IRequest, IPlayer } from "../types/types";

interface PlayerProps{
  response?:IPlayer[],
  errors?:{
    token:string
  }
} 

export const getChartStatistics = async (apiKey:string, league:string, season:string, club:string, request:IRequest<PlayerProps>)=> {
    const { url, options } = CHART_STATISTICS_API(apiKey, league, season, club)
  const {response, json} = await request(url, options)
    if(response && json){
    return {response, json}
   }
    else return 
 }