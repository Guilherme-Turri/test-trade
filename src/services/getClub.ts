import { CLUB_API } from "./api";
import { IRequest, IClub } from "../types/types";

interface ClubPros{
  response?:IClub[],
  errors?:{
    token:string
  }
} 

export const getClub = async (apiKey:string, league:string, season:string, request:IRequest<ClubPros>)=> {
  const { url, options } = CLUB_API(apiKey, league, season)
  const {response, json} = await request(url, options)
  if(response && json){
    return {response, json}
   }
    else return 
 }