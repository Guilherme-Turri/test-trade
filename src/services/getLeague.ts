import { LEAGUES_API } from "./api";
import { IRequest, ILeague } from "../types/types";

interface LeagueProps{
  response?:ILeague[],
  errors?:{
    token:string
  }
} 

export const getLeague = async (apiKey:string, country:string, season:string, request:IRequest<LeagueProps>)=> {
  const { url, options } = LEAGUES_API(apiKey, country, season)
  const {response, json} = await request(url, options)
  if(response && json){
    return {response, json}
   }
    else return 
 }