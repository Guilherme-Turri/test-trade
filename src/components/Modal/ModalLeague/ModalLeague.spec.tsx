/* eslint-disable testing-library/no-unnecessary-act */
import { pageRender } from "../../../render/pageRender";
import { fireEvent, screen, waitFor, act } from '@testing-library/react'
import { ModalLeague } from ".";
import * as reactRedux from 'react-redux'
import * as getClubTest from '../../../services/getClub';
import * as fetchDataTest from '../../../hooks/useFetch'

jest.setTimeout(10000)

const mockedUseDispatch = jest.fn()
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: () => mockedUseDispatch
}));

const allLeaguesMock = [
  {
    league: {
      id: '01',
      name: 'league01',
      logo: 'www.league.com/logo01'
    }
  },
  {
    league: {
      id: '02',
      name: 'league02',
      logo: 'www.league.com/logo02'
    }
  },
]

const parameters = {
  league: {
    all: allLeaguesMock,
  }
}
const user = {
  apiKey: '123'
}


describe('<ModalLeague/>', () => {
  const useSelectorMock = reactRedux.useSelector as jest.Mock;
  const getClubSpy = jest.spyOn(getClubTest, 'getClub');

  it('Should render on screen properly', () => {
    useSelectorMock.mockReturnValue({
      parameters
    });
    const { container } = pageRender(<ModalLeague />)
    const item = screen.getByText('league01');
    expect(item).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  })
  it('Should render on the screen with a message to redo the search.', () => {
    useSelectorMock.mockReturnValue({
      parameters: null
    });
    pageRender(<ModalLeague />)
    const item = screen.getByText('Refaça a busca e/ou verifique as requisições diárias');
    expect(item).toBeInTheDocument();
  })

  it('should dispatch selected league into context', () => {
    useSelectorMock.mockReturnValue({
      user,
      parameters
    });

    pageRender(<ModalLeague />)

    const selectedLeague = screen.getByText('league01');
    act(() => {
      fireEvent.click(selectedLeague);
    });

    expect(mockedUseDispatch).toHaveBeenCalledWith({
      "payload": {
        "league": {
          "id": "01",
          "logo": "www.league.com/logo01",
          "name": "league01"
        }
      }, "type": "parameters/setLeague"
    });

  })
  it('should fetch all clubs from api and store on context', async () => {
    useSelectorMock.mockReturnValue({
      user,
      parameters
    });

    const allClubsMock = [
      {
        team: {
          id: '01',
          name: 'Club01',
          logo: 'logo/01',
          code: '01'
        },
      },
      {
        team: {
          id: '02',
          name: 'Club02',
          logo: 'logo/02',
          code: '02'
        },
      }
    ]

    getClubSpy.mockResolvedValue({
      response: { status: 200 },
      json: {
        response: allClubsMock
      }
    });

    pageRender(<ModalLeague />)
    const selectedLeague = screen.getByText('league01');

    act(() => {
      fireEvent.click(selectedLeague);
    });

    expect(mockedUseDispatch).toHaveBeenCalledWith({
      "payload": {
        "league": {
          "id": "01",
          "logo": "www.league.com/logo01",
          "name": "league01"
        }
      }, "type": "parameters/setLeague"
    });
    await waitFor(() => {
      expect(mockedUseDispatch).toHaveBeenNthCalledWith(2, {
        "payload": allClubsMock,
        "type": "parameters/setClubs"
      });
    })
  })

  it('should display error message on getClub', async () => {
    useSelectorMock.mockReturnValue({
      user,
      parameters
    });

    getClubSpy.mockResolvedValue({
      response: { status: 200 },
      json: {
        response: [],
        errors: { token: 'Verifique novamente a busca' }
      }
    });

    pageRender(<ModalLeague />)
    const selectedLeague = screen.getByText('league01');
    act(() => {
      fireEvent.click(selectedLeague);
    });

    expect(mockedUseDispatch).toHaveBeenCalledWith({
      "payload": {
        "league": {
          "id": "01",
          "logo": "www.league.com/logo01",
          "name": "league01"
        }
      }, "type": "parameters/setLeague"
    });
    await waitFor(() => {
      const msg = screen.getByText('Verifique novamente a busca')
      expect(msg).toBeInTheDocument();
    })
  })
})

describe('<ModalLeague/> -  display right actions by the custom hook useFetch', () => {
  const useSelectorMock = reactRedux.useSelector as jest.Mock;
  it('should display loading on screen', async () => {
    jest.spyOn(fetchDataTest, "useFetchData").mockReturnValue({
      error: null,
      request: jest.fn(),
      loading: true
    });

    useSelectorMock.mockReturnValue({
      user,
      parameters
    });

    pageRender(<ModalLeague />)

    await waitFor(() => {
      expect(screen.getByTestId('Ball')).toBeInTheDocument();
    })
  })

  it('should display errormsg on screen', async () => {
    jest.spyOn(fetchDataTest, "useFetchData").mockReturnValue({
      error: 'ErrorMsg',
      request: jest.fn(),
      loading: false
    });
    useSelectorMock.mockReturnValue({
      user,
      parameters
    });

    pageRender(<ModalLeague />)
    const selectedLeague = screen.getByText('league01');
    act(() => {
      fireEvent.click(selectedLeague);
    });
    await waitFor(() => {
      expect(screen.getByText('ErrorMsg')).toBeInTheDocument();
    })
  })
})   