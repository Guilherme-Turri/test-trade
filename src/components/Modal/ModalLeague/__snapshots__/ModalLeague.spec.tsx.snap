// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalLeague/> Should render on screen properly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  width: 100%;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.938rem;
  overflow-y: auto;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-bottom: 2px solid #00c555;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 9.375rem;
  height: 3.438rem;
  margin: 0.938rem;
  cursor: pointer;
}

<div>
  <div
    class="c0"
    data-testid="modalLeagueId"
  >
    <div
      class="c1"
    >
      league01
    </div>
    <div
      class="c1"
    >
      league02
    </div>
  </div>
</div>
`;
